### ISSUE TITLE
sample() should shuffle when fraction=1

### ISSUE THREAD

**Issue by lyngc** [pola-rs/polars#13685]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
df = pl.DataFrame({"a": [1, 2, 3, 4, 5]})
df.sample(fraction=1, seed=42) # 1, 2, 3, 4, 5
df.sample(fraction=1, seed=84) # 1, 2, 3, 4, 5
df.sample(fraction=1, seed=21) # 1, 2, 3, 4, 5
```

### Issue description

When `fraction=1` `sample()` always return the data in the same order.

### Expected behavior

I would expect the data to be shuffled when `n=len(df)` or `fraction=1` in `sample()`. I know there is a `shuffle` parameter but would expect that to be `True` in this case. Else `sample()` does not do anything.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.3
Index type:           UInt32
Platform:             Windows-10-10.0.19045-SP0
Python:               3.11.1 (tags/v3.11.1:a7a450f, Dec  6 2022, 19:58:39) [MSC v.1934 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.6.3
numpy:                1.24.1
openpyxl:             <not installed>
pandas:               1.5.3
pyarrow:              11.0.0
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

**Comment by ritchie46** [comment]
This is not a bug. Polars' sample doesn't shuffle by default.

**Comment by lyngc** [comment]
But could it be an enhancement to shuffle in this case? Else, there is nothing random going on when using `sample(fraction=1)` as the result will always be the same, regardless of the seed.

**Comment by Wainberg** [comment]
yeah I'm adding a warning to the docs about this case. it should stay as-is though, to be consistent.

**Comment by lyngc** [comment]
If `sample()` does not shuffle by default and `df.sample(fraction=1, seed=42)` does not shuffle and return 1, 2, 3, 4, 5, how come `df.sample(fraction=0.8, seed=42)` return 1, 4, 3, 2 and not 1,2, 3, 4?

When faction = 0.8, it looks like a randow row was picked 4 times, which make sense. But when fraction = 1, it does not look like a random row was picked 5 times.

**Comment by JulianCologne** [comment]
> This is not a bug. Polars' sample doesn't shuffle by default.

I kind of disagree.
From a user perspective the shuffle behaviour seems unexpected at least if not inconsistent. Let me give examples:

```python
df = pl.DataFrame({"a": [1, 2, 3, 4]})

N = 3 # 4

for i in range(10):
    shuffle_true  = df.sample(n=N, shuffle=True, seed=i).get_column("a").to_list()
    shuffle_false = df.sample(n=N, shuffle=False, seed=i).get_column("a").to_list()
    print(f"Shuffle True: {shuffle_true}, shuffle False: {shuffle_false}")
```

Results when `N<len` (N=3) (always seems to shuffle)
```bash
Shuffle True: [3, 2, 4], shuffle False: [3, 2, 4]
Shuffle True: [4, 2, 1], shuffle False: [4, 2, 1]
Shuffle True: [1, 3, 2], shuffle False: [1, 3, 2]
Shuffle True: [4, 3, 1], shuffle False: [4, 3, 1]
Shuffle True: [2, 3, 4], shuffle False: [2, 3, 4]
Shuffle True: [4, 2, 3], shuffle False: [4, 2, 3]
Shuffle True: [3, 2, 4], shuffle False: [3, 2, 4]
Shuffle True: [4, 2, 1], shuffle False: [4, 2, 1]
Shuffle True: [1, 2, 4], shuffle False: [1, 2, 4]
Shuffle True: [3, 1, 2], shuffle False: [3, 1, 2]
```

Results when `N==len` (N=4) (only shuffling when shuffle=True)
```bash
Shuffle True: [4, 1, 2, 3], shuffle False: [1, 2, 3, 4]
Shuffle True: [2, 3, 4, 1], shuffle False: [1, 2, 3, 4]
Shuffle True: [3, 2, 1, 4], shuffle False: [1, 2, 3, 4]
Shuffle True: [3, 1, 2, 4], shuffle False: [1, 2, 3, 4]
Shuffle True: [4, 1, 3, 2], shuffle False: [1, 2, 3, 4]
Shuffle True: [1, 3, 4, 2], shuffle False: [1, 2, 3, 4]
Shuffle True: [3, 1, 4, 2], shuffle False: [1, 2, 3, 4]
Shuffle True: [1, 3, 4, 2], shuffle False: [1, 2, 3, 4]
Shuffle True: [3, 4, 1, 2], shuffle False: [1, 2, 3, 4]
Shuffle True: [2, 3, 4, 1], shuffle False: [1, 2, 3, 4]
```

### Confusing points:
- if n<len: parameter `shuffle` does NOT matter. Will always "shuffle" from a user perspective 
- if n=len: `shuffle` does matter.

**Comment by lyngc** [comment]
@ritchie46 Do you still believe that polars does not shuffle by default given @JulianCologne examples above? I still believe it should also shuffle then n=len

**Comment by MarcoGorelli** [comment]
This came up again in https://github.com/pola-rs/polars/issues/23250

Here's the special-casing in the code:

https://github.com/pola-rs/polars/blob/bcdc7da8599de43ecbcbfd4c04752f26343eda23/crates/polars-core/src/chunked_array/random.rs#L33-L47

As far as I can tell, `shuffle` is used _only_ when `n==len`? It doesn't appear at all in the `else` branch, nor in `create_rand_index_with_replacement`

I think the #23250 poster does have a point that https://stackoverflow.com/a/34879805/4451315 is a very popular answer, with 1610 net upvotes right now

If so, I'm inclined to agree with the request - deprecate `shuffle`, and just run `buf.shuffle` for `n == len`.

✨ This would simplify both the implementation and the public-facing API ✨

**Comment by ritchie46** [comment]
I don't think it is required to deprecate `shuffle`. I believe we do as it says:

```
If set to True, the order of the sampled rows will be shuffled. If set to False (default), the order of the returned rows will be neither stable nor fully random.
```

The else branch might require an extra shuffle (not entirely sure). We could default to `True` to better align with peoples expectations. The `shuffle=False` doesn't guarantee anything, which is what we do.

However, I am a bit hesitant while I am thinking about this. Doing that in a streaming or distributed manner is much more expensive.

I do think we need to expose a `DataFrame.shuffle`.

Maybe `n == len` branch could go though. That would still not give any guarantees, but would not special case `frac == 1`.

**Comment by deanm0000** [comment]
> Users coming from pandas generally need to know one thing...

```python
polars != pandas
```

It seems to me the fix ought to be to sort those random indices when `shuffle=False` that way there is no bifurcation between `n<len` and `n==len`. The results will always be in the order they started with unless explicitly setting `shuffle=True`.

For streaming a sample, what if it is probabilistic and approximate only? Only allow a fraction input and then, for every row, generate a uniform random number between 0 and 1. If it's lower than the fraction specified take that row otherwise discard it. With a large number of rows that should be very, but not perfectly, accurate. It only gets to be expensive if they want the shuffle which would be defaulted to False.

**Comment by strentom** [comment]
The issue https://github.com/pola-rs/polars/issues/23250 was closed as a duplicate, but in case the conversation shifts towards "not shuffling", then https://github.com/pola-rs/polars/issues/23250 isn't duplicate, because it's resolution was to highlight the difference between pandas and polars' implementation in the documentation.

> polars != pandas

It's not clear how this is relevant. In fact, Polars documentation often highlights a difference from pandas when appropriate. Just search for the word 'pandas' in polars docs.

**Comment by nameexhaustion** [comment]
From discussion:

Shuffling rows is not a streaming-friendly operation. We'd like to move to defining that the rows returned by `sample()` are always in input order, and that if shuffling is desired, to use `shuffle()` before/after the `sample()` (or if `fraction=1`, to simply use shuffle by itself.
